### PR TITLE
The Crown Reactor and other changes, Ready to review [mapping changes to all maps]

### PR DESCRIPTION
**Pull request by Bod9001**
# TODO 
* [x] ~~FIX Enumerators (on tab prefabs)~~
* [x] ~~add power component to stuff that requires it~~

# NOTES!!!
Requires scene changes for the new devices that require power to be hooked up
There might be some few accidental changes in here somewhere
Some of these features might be a bit buggy but I don't mind fixing them
Any changes needed just tell me

# Adds
* Periodic updates for update manager so you can specify how long the the delay between updatesare
* Multitool interface For setting stuff with the multitool and things like that
Made more devices use power

* netlabe now support text mesh pro
*supplying devices now can work on watts (  automatically doing the calculation to turn it into current and voltage)
* Adds radiation system (WIP, the resistivity of walls and such needs to set themselves on the meta data node still, but apart from that it seems to work well )
* Made it so underfloor and tile palette supports rotated and colour changed tiles (WIP, have to add the extra networking for it and change shaders on tile maps)
* reactor stuff
* Liquid pipes a bit WIP, (planning to add gas pipes after finished with reactor)
* Explosion manager
* Radiation does damage now
# what needs testing?
_Everything_
nah, main focus points is underfloor layer stuff and multitool changes

## Why?
Because I'm getting so far behind and holding up a few people with features that I've implemented here

so I've got a bunch of useful changes on my branch,
if I want to merge it all that would have to change would be wrapping up some of the machines I made powered
Stuff like reactors only matter if they're in the scene
I feel like the improvements have made will help With a few things
like Making a update manager that can do periodic updates with a specified time
So you don't have to do the enumerator yields stuff
Since apparently that generates long stack traces and gets slower and slower

thanks to @ThatDan123 for the finishing touches and mapping

### CODE REVIEW

**Comment by Bod9001** [comment]
should I do the scene changes in a separate PR?

**Comment by ThatDan123** [comment]
I think theres still a pog PR to come, so maybe in a different one.

**Comment by ThatDan123** [comment]
Noticed you commented out the conveyor code in the multitool, that intended?

**Comment by Bod9001** [comment]
I changed it over to use the new multitool interface

**Comment by Bod9001** [comment]
The failing tests look like They are coming from the scene changes that need to be made

File: UnityProject/Assets/Scripts/Electricity/NodeModules/ResistanceSourceModule.cs [hunk 1]
```diff
@@ -125,7 +123,7 @@ public override void PowerNetworkUpdate()
 	{
 		if (dirtyResistance)
 		{
-			if (NotEditorResistanceset)
+			if (NotEditorResistanceset )
```
Comment (daalta) [inline UnityProject/Assets/Scripts/Electricity/NodeModules/ResistanceSourceModule.cs:126]: Spess

File: UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs [hunk 2]
```diff
@@ -317,4 +317,18 @@ public void ServerPerformInteraction(HandApply interaction)
 				});
 		}
 	}
+
+	//######################################## Multitool interaction ##################################
```
Comment (daalta) [inline UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs:321]: Might as well make this a `Header["Multitool interaction"]`

**Comment by daalta** [comment]
There's a lot of changes in here. Have you checked if there's any new warnings?

Still hoping we'll get to 0 some day 🙏

File: UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs:321]: How does it work exactly?

Comment (daalta) [reply UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs:321]: Oh, headers are just text that appears in the inspector, like a headline. 

So you'll see that the following serialized variables are for "Multitool interaction", but it'll also be visible in the inspector 🙂

https://docs.unity3d.com/ScriptReference/HeaderAttribute.html

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs:321]: oh ok, thought it was some type of code separation stuff

**Comment by ThatDan123** [comment]
Are you sure putting the APC power component on scrubbers and vents is the best idea? Think it would be better to have them connect to a Air Alarm and then power is sent to that. 

Also its a right pain to remove stuff from giant lists when needed.

**Comment by Bod9001** [comment]
was Too lazy to make an air alarm system, That was before I decided to add the reactor xD

**Comment by daalta** [comment]
So there's definitely some new warnings. Trying to find a method that lets me figure our which are new.

Edit: Counting 31 new errors. Is there any way to get the warning list as a text file..?
io1qcIrN41

**Comment by Bod9001** [comment]
think there's a Unity log? somewhere in documents/app data

**Comment by daalta** [comment]
> 
> 
> think there's a Unity log? somewhere in documents/app data

Yep, found that too. But the log does not display the same warnings as the editor. I'd also have to process it to extract all warnings without timestamps...

Either way, there's some warnings in there regarding some of your changes. Most looked easy to fix.

We can either fix them up in a future PR, or we can do it here.

**Comment by Bod9001** [comment]
yeah, I don't mind sorting them out at some point

File: UnityProject/Assets/Scripts/Electricity/Engine/BoilerTurbineController.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+public class BoilerTurbineController : MonoBehaviour, ISetMultitoolSlave
+{
+
+
+	public bool State = false;
+	public ReactorBoiler ReactorBoiler = null;
+	public ReactorTurbine ReactorTurbine = null;
+    // Start is called before the first frame update
+    void Start()
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Electricity/Engine/BoilerTurbineController.cs]: Unneeded start method

File: UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs [hunk 4]
```diff
@@ -317,4 +317,19 @@ public void ServerPerformInteraction(HandApply interaction)
...
+
+	[SerializeField]
+	private MultitoolConnectionType conType = MultitoolConnectionType.Conveyor;
+	public MultitoolConnectionType ConType  => conType;
+
+	//[SerializeField]
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs]: Supposed to be commented out?

File: UnityProject/Assets/Scripts/Electricity/Engine/GUI_ReactorControler.cs [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
+
+			CoreFluxLevel.SetValueServer((Math.Round(Value).ToString()));
+			GUIReactorLayout.Refresh();
+			GUIReactorAnnunciator.Refresh();
+		}
+	}
+
+	/// <summary> As
+	/// Sets shuttle speed.
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Electricity/Engine/GUI_ReactorControler.cs]: Wrong summary??

File: UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+public class PowerControlConsole : MonoBehaviour, IAPCPowered
+{
+	public ResistanceSourceModule ThisDevice;
+	public ReactorControlConsole ControlConsole;
+	private float TargetsSupplyOutputVoltage = 275000;
+
+	private float PreviousVoltage = 0;
+
+	void IAPCPowered.PowerNetworkUpdate(float Voltage)
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs]: needed? Whole method internals commented out.

File: UnityProject/Assets/Scripts/Electricity/Engine/ReactorControlConsole.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		float ControlRodDepthPercentage = 0;
+
+		if (ReactorChambers.ControlRodDepthPercentage == 0)
+		{
+			ReactorChambers.ControlRodDepthPercentage = 0.001f;
+		}
+
+		ReactorChambers.ControlRodDepthPercentage *= (1 / Multiplier);
+		ControlRodDepthPercentage = ReactorChambers.ControlRodDepthPercentage;
+//Change this to use SetControlRodDepth
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Electricity/Engine/ReactorControlConsole.cs]: Does this need doing, also are these logs needed?

File: UnityProject/Assets/Scripts/Electricity/NodeModules/ModuleSupplyingDevice.cs [hunk 8]
```diff
@@ -69,13 +79,15 @@ public override void OnDespawnServer(DespawnInfo info)
...
+
 	public override void PowerUpdateCurrentChange()
 	{
 		PowerSupplyFunction.PowerUpdateCurrentChange(this);
 	}
-
+	[RightClickMethod]
 	public override void TurnOnSupply()
 	{
+		Logger.Log("TurnOnSupply");
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Electricity/NodeModules/ModuleSupplyingDevice.cs]: Log needed?

File: UnityProject/Assets/Scripts/Input System/InteractionV2/TileInteraction/PipeDeconstruction.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		string othersMessage = Chat.ReplacePerformer(othersStartActionMessage, interaction.Performer);
+		Chat.AddActionMsgToChat(interaction.Performer, performerStartActionMessage, othersMessage);
+		if (interaction.BasicTile.LayerType != LayerType.Underfloor) return;
+
+		var PipeTile = interaction.BasicTile as PipeTile;
+		var matrix = interaction.TileChangeManager.MetaTileMap.Layers[LayerType.Underfloor].matrix;
+		var metaDataNode = matrix.GetMetaDataNode(interaction.TargetCellPos);
+		for (var i = 0; i < metaDataNode.PipeData.Count; i++)
+		{
+			Logger.Log("OH?");
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Input System/InteractionV2/TileInteraction/PipeDeconstruction.cs]: Logs needed?

File: UnityProject/Assets/Scripts/Radiation/RadiationProducer.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+			}
+
+			UpdateValues(OutPuttingRadiation);
+
+			//Logger.Log("RequestPulse!!" + Time.time);
+		}
+
+		// Update is called once per frame
+		void Update()
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Radiation/RadiationProducer.cs]: empty update method

File: UnityProject/Assets/Scripts/Radiation/RadiationProducer.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+			mLightRendererObject = LightSpriteBuilder.BuildDefault(gameObject, color, 7);
+			mLightRendererObject.SetActive(true);
+
+			lightSprite = mLightRendererObject.GetComponent<LightSprite>();
+			lightSprite.Sprite = DotSprite;
+			registerObject = this.GetComponent<RegisterObject>();
+		}
+
+		// Start is called before the first frame update
+		void Start()
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Radiation/RadiationProducer.cs]: empty start method

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/MetaDataSystem.cs [hunk 12]
```diff
@@ -211,11 +211,11 @@ private void SetupNeighbors(MetaDataNode node)
 
 			if (metaTileMap.IsSpaceAt(neighbor, true))
 			{
-				// if current node is a room, but the neighboring is a space tile, this node needs to be checked regularly for changes by other matrices
-				if (node.IsRoom && !externalNodes.ContainsKey(node))
+				/*// if current node is a room, but the neighboring is a space tile, this node needs to be checked regularly for changes by other matrices
+				if (node.IsRoom && !externalNodes.ContainsKey(node) && metaTileMap.IsSpaceAt(node.Position, true) == false)
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/MetaDataSystem.cs:215]: This supposed to be commented out?

File: UnityProject/Assets/Scripts/UI/Net/Elements/NetSlider.cs [hunk 13]
```diff
@@ -24,7 +24,7 @@ public override string Value
 
 	public FloatEvent ServerMethod;
 
-	private Slider element;
+	public Slider element;
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/UI/Net/Elements/NetSlider.cs]: Needs to be public?

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/MetaDataSystem.cs (hunk 12 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/MetaDataSystem.cs:215]: I was having lag issues with the explosion, Turns out if I commenting this out Solve the lag issue And it seems to not impact anything

File: UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs (hunk 6 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs]:
```
{
		/*ElectricityFunctions.WorkOutActualNumbers(ThisDevice.ControllingNode.Node.InData);
		var ControllerMultiplier = (TargetsSupplyOutputVoltage+1) / (ThisDevice.ControllingNode.Node.InData.Data.ActualVoltage+1);
		Logger.Log("in ControllerMultiplier " + ControllerMultiplier+"  >> " +  (TargetsSupplyOutputVoltage+1)+ " / " + (ThisDevice.ControllingNode.Node.InData.Data.ActualVoltage+1));
		if (ControllerMultiplier != 0 && !float.IsInfinity(ControllerMultiplier))
		{
			ControlConsole.RequestRelativeChange(ControllerMultiplier);
		}
		PreviousVoltage = ThisDevice.ControllingNode.Node.InData.Data.ActualVoltage;*/

	}
```

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs]: yeah, It's not used for anything currently was meant for automatically increasing the output of the reactor but is a bit more tricky to calculate then I realised Should I remove it?

File: UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Construction/ConveyorBelts/ConveyorBelt.cs]: yeah Unity was being Dumdum with Changing the values in the code but not updating on the serialised inspector

File: UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs (hunk 6 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Electricity/Engine/PowerControlConsole.cs]: Guess it depends on whether you want to come back and try again, no harm keeping it in if thats the case, though maybe add a comment on it just so you dont forget why its commented out :p

File: UnityProject/Assets/Scripts/Electricity/Electrical processes/ElectricalSynchronisation.cs [hunk 14]
```diff
@@ -380,7 +382,7 @@ private void PowerUpdateResistanceChange()
 
 		for (int i = ResistanceChange.Count - 1; i >= 0; i--)
 		{
-			if (i < InitialiseResistanceChange.Count)
+			if (i < ResistanceChange.Count)
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Electricity/Electrical processes/ElectricalSynchronisation.cs]: redundant

**Comment by Bod9001** [comment]
any more changes that are needed?
apart from sprites

**Comment by ThatDan123** [comment]
Only problem ive found is that theres a random vendor spawner for maps but they cant be connected to the APCs, this is fine for now as they are non essential and theres like one of each map, so not urgent.

**Comment by ThatDan123** [comment]
Bod can you rename this to all maps are being changed.

**Comment by Bod9001** [comment]
merging because other people need to do mapping

**Pull request merged**